controllers/secretagentconfiguration_controller.go

```diff
@@ -130,210 +134,74 @@ func (reconciler *SecretAgentConfigurationReconciler) Reconcile(req ctrl.Request
-					}
-				}
-			}
-			keyInterface.ToKubernetes(secObject)
-			// if we reach this point, we need to update the k8s secret with new keys
-			k8sApplyRequired = true
-		} // end of keys loop
-		// if redoKeys is true, rerun the keys loop
-		if redoKeys {
-			log.V(1).Info("retrying all keys from secret",
-				"secret_name", secretReq.Name)
-			redoKeys = false
-			// HOTFIX: This is bad practice. We shoudln't use `goto`. The reconcile loop must be refactored asap.
-			goto secretKeysLoop
+		gen := generator.GenConfig{
+			// kubernetes secret that will have keys
+			SecObject: secObject,
+			// Keys that should be in secret
+			KeysToGen: secretReq.Keys,
+			Log:       log,
+			Namespace: instance.Namespace,
+			AppConfig: &instance.Spec.AppConfig,
+			Client:    reconciler.Client,
+			// TODO this is for secret manager refactor in the near future
+			// SecretManagerClient: secretMgr,
 		}
-		// write secret once after processing all its keys if required.
-		if k8sApplyRequired {
-			log.V(0).Info("applying to kubernetes",
-				"secret_name", secretReq.Name)
-
-			secObject.Labels = labelsForSecretAgent(instance.Name)
-			// Set SecretAgentConfiguration instance as the owner and controller of the k8ssecret
-			if err := ctrl.SetControllerReference(&instance, secObject, reconciler.Scheme); err != nil {
-				return ctrl.Result{}, err
-			}
-			op, err := k8ssecrets.ApplySecrets(reconciler.Client, secObject)
-			if err != nil {
-				log.Error(err, "couldnt apply secret",
-					"method", op,
-					"secret_namespace", instance.Namespace,
-					"secret_name", secretReq.Name)
-				rescheduleRetry, errorFound = true, true
-			}
-			updatedK8sSecrets = true
-			// Give enough time for the k8s api to update the secret to avoid race conditions
-			time.Sleep(100 * time.Millisecond)
+		// generate this secrets keys
+		err = gen.GenKeys(ctx)
+		if err != nil {
+			// report err and retry
+			rescheduleRetry = true
+			continue
+		}
+		log.V(0).Info("applying to kubernetes")
+
+		secObject.Labels = labelsForSecretAgent(instance.Name)
+		// Set SecretAgentConfiguration instance as the owner and controller of the k8ssecret
+		if err := ctrl.SetControllerReference(&instance, secObject, reconciler.Scheme); err != nil {
+			// log error
+			rescheduleRetry = true
+			continue
 		}
+		op, err := k8ssecrets.ApplySecrets(reconciler.Client, secObject)
+		if err != nil {
+			log.Error(err, "couldnt apply secret",
+				"method", op)
+			rescheduleRetry, errorFound = true, true
+			continue
 
-		log.V(1).Info("completed reconcile for secret",
-			"secret_namespace", instance.Namespace,
-			"secret_name", secretReq.Name)
+		}
+	}
 
-	} // end of secrets loop
 	// Only update the instance's status if there was a k8s operation.
 	// No k8s operation happen in the last reconcile loop. Update to "completed" if no updates and no errors occurred
-	if updatedK8sSecrets || rescheduleRetry || errorFound {
+	if rescheduleRetry || errorFound {
```

Comment:
You'll want to keep the `updatedK8sSecrets` variable. if updatedK8sSecrets == True and `rescheduleRetry = errorFound = false` That means you just wrote the last secret or set of secrets. That's when we update the status to `Completed`